File: 20. 有效的括号/main.go
package main

import "fmt"

func isValid(s string) bool {
	l := len(s)
	if l == 1 {
		return false
	}
	for i := 0; i < l; i += 2 {
		switch string(s[i]) {
		case "(":
			if string(s[i+1]) != ")" && string(s[l-i-1]) != ")" {
				return false
			}
		case "{":
			if string(s[i+1]) != "}" && string(s[l-i-1]) != "}" {
				return false
			}
		case "[":
			if string(s[i+1]) != "]" && string(s[l-i]-1) != "]" {
				return false
			}
		}
	}

	return true
}

func main() {
	s := "{[]}"
	fmt.Println(isValid(s))
}

File: 20. 有效的括号/main.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  let len = s.length
  if (len % 2 !== 0) {
    return false
  }

  const pair = new Map([
    [')', '('],
    ['}', '{'],
    [']', '['],
  ])

  const stak = []

  for (let i = 0; i < len; i++) {
    if (pair.has(s[i]) && stak.length) {
        if (stak[stak.length - 1] !== pair.get(s[i])) {
          return false
        }
        stak.pop()
     
    }else {
      stak.push(s[i])
    }
    console.log(stak)
  }
  return !stak.length
}

let s="))"
console.log(isValid(s))
